### PR TITLE
feat: allow ordering to be config based in addition to folder based

### PR DESCRIPTION
**Pull request by Nemikolh**
This PR makes it possible to specify the ordering in the metadata of the `tutorial`, `part` or `chapter`. If it isn't specified it fallback to the current behaviour and when it's used it prints warning if it finds any issues:

image

## How to specify the order

In `tutorial/meta.md`:

```yaml
type: tutorial
parts:
  - foo         # Must match the name of the folder containing the part
  - bar
```

In `tutorial/foo/meta.md`:

```yaml
type: part
chapters:
  - chapterfoo  # Must match the name of the folder containing the chapter
```

In `tutorial/foo/chapterfoo/meta.md`:

```yaml
type: chapter
lessons:
  - welcome    # Must match the name of the folder containing the lesson
```

### CODE REVIEW

File: packages/astro/src/default/pages/index.astro [hunk 1]
```diff
@@ -3,10 +3,9 @@ import { getTutorial } from '../utils/content';
 
 const tutorial = await getTutorial();
 
-const parts = Object.values(tutorial.parts);
-const part = parts[0];
-const chapter = part.chapters[1];
-const lesson = chapter.lessons[1];
```
Comment (SamVerschueren) [inline packages/astro/src/default/pages/index.astro:9]: Why was this 1-based?

File: packages/astro/src/default/pages/index.astro [hunk 2]
```diff
@@ -3,10 +3,9 @@ import { getTutorial } from '../utils/content';
...
+const part = tutorial.parts[tutorial.firstPartId!];
+const chapter = part.chapters[part?.firstChapterId!];
+const lesson = chapter.lessons[chapter?.firstLessonId!];
```
Comment (SamVerschueren) [inline packages/astro/src/default/pages/index.astro:8]: I'm very confused. So all the `firstPartId` and `firstChapterId` and `firstLessonId` are defined as optional in `entities/index.ts`. Why are we here telling TS that they will be defined?

Also, what if they are defined but they don't exist? Then `chapter` will be undefined and it will throw, no?

**Review by SamVerschueren — COMMENTED** [review]
It's a bit hard to follow for me, but maybe it is because I don't really know how it worked before (except that it was based on the folder name).

File: packages/astro/src/default/utils/content.ts [hunk 3]
```diff
@@ -180,32 +304,15 @@ function sortCollection(collection: CollectionEntryTutorial[]) {
...
-      const numB = parseInt(splitB[i], 10);
-
-      if (!isNaN(numA) && !isNaN(numB) && numA !== numB) {
-        return numA - numB;
-      } else {
-        if (splitA[i] !== splitB[i]) {
-          return splitA[i].localeCompare(splitB[i]);
-        }
+      if (splitA[i] !== splitB[i]) {
+        return splitA[i].localeCompare(splitB[i]);
```
[the remark this answers is not in the document]
Comment (Nemikolh) [reply packages/astro/src/default/utils/content.ts:308]: Actually I'll remove that entire for loop. It's not needed anymore and it's confusing.

File: packages/astro/src/default/pages/index.astro (hunk 1 above)
Comment (Nemikolh) [reply packages/astro/src/default/pages/index.astro:9]: The folder were expected to start at `1`. We now instead pick the first folder that comes up. It could start at anything: 0, 4, ... or be explicitly defined. 

So this is a new tiny feature! :smiley:

File: packages/astro/src/default/utils/content.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (SamVerschueren) [reply packages/astro/src/default/utils/content.ts]: So the current behaviour still works then?

**Review by SamVerschueren — APPROVED** [review]

**Review by SamVerschueren — APPROVED** [review]
Great work!
